ipa-core/src/protocol/hybrid/oprf.rs

```diff
@@ -0,0 +1,313 @@
+    prf_of_match_key: u64,
+    value: Replicated<V>,
+    breakdown_key: Replicated<BK>,
+}
+
+#[tracing::instrument(name = "compute_prf_for_inputs", skip_all)]
+pub async fn compute_prf_for_inputs<C, BK, V>(
+    ctx: C,
+    input_rows: &[IndistinguishableHybridReport<BK, V>],
+) -> Result<Vec<PRFIndistinguishableHybridReport<BK, V>>, Error>
+where
+    C: UpgradableContext + ShardedContext,
+    BK: BooleanArray,
+    V: BooleanArray,
+    Replicated<Boolean, CONV_CHUNK>: BooleanProtocols<DZKPUpgraded<C>, CONV_CHUNK>,
+    Replicated<Fp25519, PRF_CHUNK>:
+        PrfSharing<MacUpgraded<C, Fp25519>, PRF_CHUNK, Field = Fp25519> + FromPrss,
+    Replicated<RP25519, PRF_CHUNK>:
+        Reveal<MacUpgraded<C, Fp25519>, Output = <RP25519 as Vectorizable<PRF_CHUNK>>::Array>,
+{
+    let conv_records =
+        TotalRecords::specified(div_round_up(input_rows.len(), Const::<CONV_CHUNK>))?;
+    let eval_records = TotalRecords::specified(div_round_up(input_rows.len(), Const::<PRF_CHUNK>))?;
+    let convert_ctx = ctx.set_total_records(conv_records);
+
+    let validator = convert_ctx.dzkp_validator(
+        MaliciousProtocolSteps {
+            protocol: &HybridStep::ConvertFp25519,
+            validate: &HybridStep::ConvertFp25519Validate,
+        },
+        CONV_PROOF_CHUNK,
+    );
+    let m_ctx = validator.context();
+
+    let curve_pts = seq_join(
+        ctx.active_work(),
+        process_slice_by_chunks(input_rows, move |idx, records: ChunkData<_, CONV_CHUNK>| {
+            let record_id = RecordId::from(idx);
+            let input_match_keys: &dyn Fn(usize) -> Replicated<MatchKey> =
+                &|i| records[i].match_key.clone();
+            let match_keys =
+                BitDecomposed::<Replicated<Boolean, 256>>::transposed_from(input_match_keys)
+                    .unwrap_infallible();
+            convert_to_fp25519::<_, CONV_CHUNK, PRF_CHUNK>(m_ctx.clone(), record_id, match_keys)
+        }),
+    )
+    .map_ok(Chunk::unpack::<PRF_CHUNK>)
+    .try_flatten_iters()
+    .try_collect::<Vec<_>>()
+    .await?;
+
+    // In order to compute the same PRF value for the same match key
+    // all shards must have the same prf_key. This has the leader
+    // generate a key, and give it to all the followers.
+    let prf_key_ctx = ctx.set_total_records(TotalRecords::ONE);
+    let prf_key = if ctx.shard_id() == ShardIndex::FIRST {
+        let leader_prf_key = gen_prf_key(&ctx.narrow(&HybridStep::PrfKeyGen));
+
+        for shard in ctx.shard_count().iter().skip(1) {
+            prf_key_ctx
+                .clone()
+                .shard_send_channel::<Replicated<Fp25519>>(shard)
+                .send(RecordId::FIRST, leader_prf_key.clone())
+                .await?;
+        }
+        Ok(leader_prf_key)
+    } else {
+        let mut receiver = prf_key_ctx
+            .shard_recv_channel::<Replicated<Fp25519>>(ShardIndex::FIRST)
+            .take(1);
+        receiver.next().await.unwrap()
+    }
+    .unwrap();
```

Comment:
I couldn't seem to get this to work with `?` instead of the `unwrap()`s here, which is not great. any suggestions on this would be helpful.